src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -1122,15 +1123,23 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
      * Returns a list of trees that each should be concatenated, from left to right.
      * It turns a chained call like "a".+("b").+("c") into a list of arguments.
      */
-    def liftStringConcat(tree: Tree): List[Tree] = tree match {
-      case Apply(fun @ Select(larg, method), rarg) =>
-        if (isPrimitive(fun.symbol) &&
-            scalaPrimitives.getPrimitive(fun.symbol) == scalaPrimitives.CONCAT)
-          liftStringConcat(larg) ::: rarg
-        else
-          tree :: Nil
-      case _ =>
-        tree :: Nil
+    def liftStringConcat(tree: Tree): List[Tree] = {
+      val result = ListBuffer[Tree]()
+      def loop(tree: Tree): Unit = {
+        tree match {
+          case Apply(fun@Select(larg, method), rarg :: Nil) =>
+            if (isPrimitive(fun.symbol) &&
+              scalaPrimitives.getPrimitive(fun.symbol) == scalaPrimitives.CONCAT) {
+              loop(larg)
+              loop(rarg)
```

Comment:
@sjrd IIRC ScalaJS can just leave optimization of string concatenation to the Javascript VM, so you don't have analagous code in your backend to modify. Would the VMs would optimize `(a + b + c + d)` and `((a + b) + (c + d))` identically?

Reply:
Actually we already have code in our optimizer to normalize `x + (y + z)` into `(x + y) + z`:
https://github.com/scala-js/scala-js/blob/master/linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala#L3097-L3098

That means that we will already rewrite `((a + b) + (c + d))` into `(a + b + c + d)`.